executor/aggregate_xapi.go

```diff
@@ -0,0 +1,175 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package executor
+
+import (
+	"strings"
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/util/types"
+)
+
+var (
+	_ Executor = &XAggregateExec{}
+)
+
+// finalAggregater is used to merge all partial aggregate results from multiple regions and get final result.
+type finalAggregater struct {
+	name         string
+	currentGroup []byte
+	// contextPerGroupMap is used to store aggregate evaluation context.
+	// Each entry for a group.
+	contextPerGroupMap map[string](*ast.AggEvaluateContext)
+}
+
+// Update is used for update aggregate context.
+func (n *finalAggregater) update(count uint64, value types.Datum) error {
+	switch n.name {
+	case ast.AggFuncCount:
+		return n.updateCount(count)
+	}
+	return nil
+}
+
+// GetContext gets aggregate evaluation context for the current group.
+// If it is nil, add a new context into contextPerGroupMap.
+func (n *finalAggregater) getContext() *ast.AggEvaluateContext {
+	if n.contextPerGroupMap == nil {
+		n.contextPerGroupMap = make(map[string](*ast.AggEvaluateContext))
+	}
+	if _, ok := n.contextPerGroupMap[string(n.currentGroup)]; !ok {
+		n.contextPerGroupMap[string(n.currentGroup)] = &ast.AggEvaluateContext{}
+	}
+	return n.contextPerGroupMap[string(n.currentGroup)]
+}
+
+func (n *finalAggregater) updateCount(count uint64) error {
+	ctx := n.getContext()
+	ctx.Count += int64(count)
+	return nil
+}
+
+// XAggregateExec deals with all the aggregate functions.
+// It is built from Aggregate Plan. When Next() is called, it reads all the data from Src and updates all the items in AggFuncs.
+// TODO: Support having.
+type XAggregateExec struct {
+	Src               Executor
+	ResultFields      []*ast.ResultField
+	AggFields         []*types.FieldType
+	executed          bool
+	ctx               context.Context
+	AggFuncs          []*ast.AggregateFuncExpr
+	aggregaters       []*finalAggregater
+	groupMap          map[string]bool
+	groups            []string
+	currentGroupIndex int
+}
+
+// Fields implements Executor Fields interface.
+func (e *XAggregateExec) Fields() []*ast.ResultField {
+	return e.ResultFields
+}
+
+// Next implements Executor Next interface.
+func (e *XAggregateExec) Next() (*Row, error) {
+	// In this stage we consider all data from src as a single group.
+	if !e.executed {
+		e.groupMap = make(map[string]bool)
+		e.groups = []string{}
+		e.aggregaters = make([]*finalAggregater, len(e.AggFuncs))
+		for i, af := range e.AggFuncs {
+			agg := &finalAggregater{
+				name: strings.ToLower(af.F),
+			}
+			e.aggregaters[i] = agg
+		}
+		for {
+			hasMore, err := e.innerNext()
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			if !hasMore {
+				break
+			}
+		}
+		e.executed = true
+		if len(e.groups) == 0 {
```

Comment:
Seems like this branch will never be executed, as each local region returns single groupKey if no group by items.